src/rtc.c

```diff
@@ -0,0 +1,518 @@
+ *
+ * To check if the real-time clock is available, call #joybus_rtc_is_present.
+ * If detected, it is a good idea to ensure the RTC control block is in a
+ * known-good state by calling the #joybus_rtc_run function before reading
+ * the current time with #joybus_rtc_read_time for the first time.
+ *
+ * The simplest method to change the time is by calling #joybus_rtc_set,
+ * which is a high-level convenience function that prepares the RTC and
+ * writes the time with the proper delays.
+ *
+ * In general, since only one game ever used Joybus RTC (and that game was
+ * later re-released on the GameCube in English), RTC support by emulators
+ * and flash carts tends to be incomplete, inaccurate, or non-existent.
+ * Many emulators do not actually implement the RTC write command and will
+ * always respond with the host system's current local time. Some emulators
+ * and flash carts support writing to RTC but will not persist the date/time
+ * after resetting or powering-off.
+ *
+ * The only way to check if writes are actually supported is to write a
+ * time to the RTC and read the time back out. Many emulators that do
+ * support RTC reads will silently ignore RTC writes. You should detect
+ * whether writes are supported using #joybus_rtc_is_writable so that you can
+ * conditionally show the option to change the time if it's supported.
+ *
+ * The RTC sends and receives the date/time in bytes of binary-coded decimal
+ * numbers. This subsystem handles decoding and encoding the date/time into
+ * a more generally-useful struct called #rtc_time_t.
+ *
+ * Special thanks to marshallh and jago85 for their hard work and research
+ * reverse-engineering and documenting the inner-workings of the Joybus RTC.
+ *
+ * @{
+ */
+
+/**
+ * @brief Real-time clock identifier byte.
+ *
+ * A magic value in the RTC status response that indicates the RTC is present.
+ */
+#define JOYBUS_RTC_IDENTIFIER 0x10
+
+/**
+ * @brief Duration (in milliseconds) to wait after writing to RTC.
+ *
+ * The software must wait for the previous RTC write to finish before issuing
+ * another RTC write command. Since there is no status returned by the RTC to
+ * indicate that a write has completed, the software must wait a fixed duration.
+ */
+#define JOYBUS_RTC_WRITE_DELAY 20
+
+/**
+ * @brief The RTC is ready for block 2 to be read.
+ */
+#define JOYBUS_RTC_STATUS_READY 0x00
+/**
+ * @brief The RTC is ready for block 2 to be written.
+ */
+#define JOYBUS_RTC_STATUS_BUSY  0x80
+
+/**
+ * @brief Control mode for setting the RTC date/time.
+ *
+ * Clears the write-protection bits and sets the busy bit.
+ */
+#define JOYBUS_RTC_CONTROL_MODE_SET 0x0004
+/**
+ * @brief Control mode for normal operation of the RTC.
+ *
+ * Clears the busy bit sets the write-protection bits.
+ *
+ * 0x0100 is the block 1 write-protection bit.
+ * 0x0200 is the block 2 write-protection bit.
+ *
+ * RTC block 1 has an unknown purpose and is typically unimplemented.
```

Comment:
Unimplemented by whom? Emulators? I'd mention it explicitly.

Do we know the actual RTC chip used by animal crossing? Its data sheet might provide some insights even di the communication is shielded by the PIF

Reply:
I have not found any documentation about what block 1 is for. No emulator implements it (that I have found), the UltraPIF and 64drive also do not implement it.

As far as I can tell, Animal Forest does not ever access block 1, so its purpose is a complete mystery. I'll do a bit more research and see if I can figure out what it is from the actual component.

Reply:
If I remember correctly, block 1 is an 8 byte battery backed SRAM.
I could do some tests if needed. I have a real RTC and with the UltraPIF and a debugger I can directly send commands to the RTC and see the results.

Reply:
The actual component does not appear to be an off-the-shelf RTC; it's branded `Nintendo RTCK-NUS` and `A0108C`. Not surprisingly, I am not able to find a datasheet based on that. 
![animalforestrtc](https://user-images.githubusercontent.com/11916110/126412286-4283b634-920f-43cf-9269-bd399386769f.jpg)